patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,5591 @@
++import com.google.common.graph.GraphBuilder;
++import com.google.common.graph.Graphs;
++import com.google.common.graph.MutableGraph;
++import io.papermc.paper.plugin.configuration.PluginConfiguration;
++import io.papermc.paper.plugin.provider.PluginProvider;
++import io.papermc.paper.plugin.provider.configuration.PaperPluginConfiguration;
++import io.papermc.paper.plugin.provider.configuration.type.PaperCommand;
++import io.papermc.paper.plugin.provider.configuration.type.PaperCommandMap;
++import io.papermc.paper.plugin.provider.service.classloader.ConfiguredPluginClassLoader;
++import io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoaderStorage;
++import io.papermc.paper.plugin.provider.service.entrypoint.Entrypoint;
++import io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.strategy.PluginGraphCycleException;
++import io.papermc.paper.plugin.provider.source.impl.DirectoryProviderSource;
++import io.papermc.paper.plugin.provider.source.impl.FileProviderSource;
++import org.bukkit.Bukkit;
++import org.bukkit.Server;
++import org.bukkit.World;
++import org.bukkit.command.Command;
++import org.bukkit.command.CommandMap;
++import org.bukkit.command.PluginCommand;
++import org.bukkit.command.PluginCommandYamlParser;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.server.PluginDisableEvent;
++import org.bukkit.event.server.PluginEnableEvent;
++import org.bukkit.plugin.*;
++import org.bukkit.plugin.java.JavaPlugin;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++import org.spongepowered.configurate.serialize.SerializationException;
++
++import java.io.IOException;
++import java.nio.file.Files;
++import java.nio.file.Path;
++import java.util.*;
++import java.util.logging.Level;
++import java.util.logging.Logger;
++
++class PaperPluginInstanceManager {
++
++    private static final FileProviderSource FILE_PROVIDER_SOURCE = new FileProviderSource(false);
++    private static final DirectoryProviderSource DIRECTORY_PROVIDER_SOURCE = new DirectoryProviderSource();
++
++    private final List<Plugin> plugins = new ArrayList<>();
++    private final Map<String, Plugin> lookupNames = new HashMap<>();
++
++    private final PluginManager pluginManager;
++    private final CommandMap commandMap;
++    private final Server server;
++
++    private final MutableGraph<String> dependencyGraph = GraphBuilder.directed().build();
++
++    public PaperPluginInstanceManager(PluginManager pluginManager, CommandMap commandMap, Server server) {
++        this.commandMap = commandMap;
++        this.server = server;
++        this.pluginManager = pluginManager;
++    }
++
++    public @Nullable Plugin getPlugin(@NotNull String name) {
++        return this.lookupNames.get(name.replace(' ', '_').toLowerCase(java.util.Locale.ENGLISH)); // Paper
++    }
++
++    public @NotNull Plugin[] getPlugins() {
++        return this.plugins.toArray(new Plugin[0]);
++    }
```

Comment:
I think it'll be better to make it return a List<Plugins>, and only in PaperPluginManagerImpl make the toArray operation.

Reply:
This is the API implementation, I can't really change this and there is no reason to. I'm not really wanting to expand the api like this.

Reply:
nvm, didn't see it was a server patch...